File: algorithms/2_medium/D/00707_design_linked_list/doubly_linked_list_smart_ptrs.cpp
class MyLinkedList {
    struct Node {
        int val;
        shared_ptr<Node> next;
        shared_ptr<Node> prev;

        Node(int value) : val{ value }, next{ nullptr }, prev{ nullptr } {}
    };

    size_t len;
    shared_ptr<Node> head;
    shared_ptr<Node> tail;

    inline shared_ptr<Node> get_kth_from_begin(const size_t k) const {
        auto curr{ head };
        for (size_t i{ 0 }; i < k; ++i) {
            curr = curr->next;
        }
        return curr;
    }

    inline shared_ptr<Node> get_kth_from_end(const size_t k) const {
        auto curr{ tail };
        for (size_t i{ 0 }; i < k; ++i) {
            curr = curr->prev;
        }
        return curr;
    }
public:
    MyLinkedList() : len{ 0 }, head{ nullptr }, tail{ nullptr } {}

    int get(const int index) const {
        const size_t from_left{ static_cast<size_t>(index) };
        if (from_left >= len) { return -1; }
        const size_t from_right{ len - 1 - from_left };
        const auto curr{ from_left <= from_right ? get_kth_from_begin(from_left) : get_kth_from_end(from_right) };
        return curr->val;
    }

    void addAtHead(int val) {
        auto node = make_shared<Node>(val);

        if (!head) {
            head = node;
            tail = node;
        } else {
            head->prev = node;
            node->next = head;
            head = node;
        }

        len += 1;
    }

    void addAtTail(int val) {
        auto node = make_shared<Node>(val);

        if (!tail) {
            head = node;
            tail = node;
        } else {
            tail->next = node;
            node->prev = tail;
            tail = node;
        }

        len += 1;
    }

    void addAtIndex(int index, int val) {
        const size_t from_left{ static_cast<size_t>(index) };
        if (from_left > len) { return; }
        const size_t from_right{ len - 1 - from_left };

        if (from_left == 0) {
            addAtHead(val);
        } else if (from_left == len) {
            addAtTail(val);
        } else {
            auto curr{ from_left <= from_right ? get_kth_from_begin(from_left) : get_kth_from_end(from_right) };
            auto node{ make_shared<Node>(val) };
            node->next = curr;
            node->prev = curr->prev;
            node->prev->next = node;
            curr->prev = node;

            len += 1;
        }
    }

    void deleteAtIndex(const int index) {
        const size_t from_left{ static_cast<size_t>(index) };
        if (from_left >= len) { return; }
        const size_t from_right{ len - 1 - from_left };

        if (len == 1) {
            head = nullptr;
            tail = nullptr;
            len = 0;
        } else if (from_left == 0) {
            head = head->next;
            head->prev = nullptr;
            len -= 1;
        } else if (from_right == 0) {
            tail = tail->prev;
            tail->next = nullptr;
            len -= 1;
        } else {
            auto curr{ from_left <= from_right ? get_kth_from_begin(from_left) : get_kth_from_end(from_right) };
            auto next{ curr->next };
            auto prev{ curr->prev };
            next->prev = prev;
            prev->next = next;
            len -= 1;
        }
    }
};

File: algorithms/2_medium/D/00707_design_linked_list/doubly_linked_list_smart_ptrs.rs
use std::cell::RefCell;
use std::rc::Rc;

struct Node {
    val: i32,
    next: Option<Rc<RefCell<Node>>>,
    prev: Option<Rc<RefCell<Node>>>,
}

impl Node {
    fn new(val: i32) -> Self {
        Self {
            val: val,
            next: None,
            prev: None,
        }
    }
}

struct MyLinkedList {
    len: usize,
    head: Option<Rc<RefCell<Node>>>,
    tail: Option<Rc<RefCell<Node>>>,
}

impl MyLinkedList {
    #[inline(always)]
    fn get_kth_from_begin(&self, k: usize) -> Rc<RefCell<Node>> {
        let mut curr = self.head.as_ref().unwrap().clone();
        for i in 0..k {
            let next = curr.borrow().next.as_ref().unwrap().clone();
            curr = next;
        }
        curr
    }

    #[inline(always)]
    fn get_kth_from_end(&self, k: usize) -> Rc<RefCell<Node>> {
        let mut curr = self.tail.as_ref().unwrap().clone();
        for i in 0..k {
            let next = curr.borrow().prev.as_ref().unwrap().clone();
            curr = next;
        }
        curr
    }

    fn new() -> Self {
        Self {
            len: 0,
            head: None,
            tail: None,
        }
    }

    fn get(&self, index: i32) -> i32 {
        let from_left = index as usize;
        if (from_left >= self.len) { return -1; }
        let from_right = self.len - 1 - from_left;
        let curr = if from_left <= from_right { self.get_kth_from_begin(from_left) } else { self.get_kth_from_end(from_right) };
        curr.borrow().val
    }

    fn add_at_head(&mut self, val: i32) {
        let mut node = Rc::new(RefCell::new(Node::new(val)));

        match self.head.clone() {
            Some(head_node) => {
                head_node.borrow_mut().prev = Some(node.clone());
                node.borrow_mut().next = Some(head_node);
                self.head = Some(node);
            },
            None => {
                self.head = Some(node.clone());
                self.tail = Some(node);
            }
        }

        self.len += 1;
    }

    fn add_at_tail(&mut self, val: i32) {
        let mut node = Rc::new(RefCell::new(Node::new(val)));

        match self.tail.clone() {
            Some(tail_node) => {
                tail_node.borrow_mut().next = Some(node.clone());
                node.borrow_mut().prev = Some(tail_node);
                self.tail = Some(node);
            },
            None => {
                self.head = Some(node.clone());
                self.tail = Some(node);
            }
        }

        self.len += 1;
    }

    fn add_at_index(&mut self, index: i32, val: i32) {
        let from_left = index as usize;
        if (from_left > self.len) { return; }
        let from_right = self.len - 1 - from_left;

        if from_left == 0 {
            self.add_at_head(val);
        } else if from_left == self.len {
            self.add_at_tail(val);
        } else {
            let mut curr = if from_left <= from_right { self.get_kth_from_begin(from_left) } else { self.get_kth_from_end(from_right) };
            let mut node = Rc::new(RefCell::new(Node::new(val)));
            node.borrow_mut().next = Some(curr.clone());
            node.borrow_mut().prev = curr.borrow_mut().prev.take();
            node.borrow_mut().prev.as_ref().unwrap().borrow_mut().next = Some(node.clone());
            curr.borrow_mut().prev = Some(node);

            self.len += 1;
        }
    }

    fn delete_at_index(&mut self, index: i32) {
        let from_left = index as usize;
        if from_left >= self.len { return; }
        let from_right = self.len - 1 - from_left;

        if self.len == 1 {
            self.head = None;
            self.tail = None;
            self.len = 0;
        } else if from_left == 0 {
            let next = self.head.as_mut().unwrap().borrow_mut().next.take();
            self.head = next;
            self.head.as_mut().unwrap().borrow_mut().prev = None;
            self.len -= 1;
        } else if (from_right == 0) {
            let prev = self.tail.as_mut().unwrap().borrow_mut().prev.take();
            self.tail = prev;
            self.tail.as_mut().unwrap().borrow_mut().next = None;
            self.len -= 1;
        } else {
            let mut curr = if from_left <= from_right { self.get_kth_from_begin(from_left) } else { self.get_kth_from_end(from_right) };
            let mut next = curr.borrow_mut().next.take();
            let mut prev = curr.borrow_mut().prev.take();
            next.as_mut().unwrap().borrow_mut().prev = prev.clone();
            prev.as_mut().unwrap().borrow_mut().next = next;
            self.len -= 1;
        }
    }
}
